### ISSUE TITLE
Possibility to change protocol in Interceptor on redirect

### ISSUE THREAD

**Issue by anycz** [square/okhttp#5515]
We have a usecase where TLS/mTLS is handled by a sidecar component (in our case, istio) and our requests need to be done using plain HTTP from applicative side (trying to call using HTTPS will make the call fail).
Some of the urls to call aren't handled by our organization and can have redirects to other https urls, hence have to be handled from applicative side.

We'd like to override protocol from https to http for redirects, but it seems it is not possible at the moment. In RealCall.getResponseWithInterceptorChain:
* the order of core internal Interceptors is fixed
* client.interceptors are before RetryAndFollowUpInterceptor and thus do not modify redirect requests
* client.networkInterceptors are after ConnectInterceptor and thus the call fails with "javax.net.ssl.SSLException: Unrecognized SSL message, plaintext connection?"
* visibility level is internal, hence extension is not possible to override getResponseWithInterceptorChain

Only current workaround found is to disable client.followRedirects and copy/paste part of RetryAndFollowUpInterceptor, but that is not maintainable.

Would it be possible to allow extra interceptors to be inserted between RetryAndFollowUpInterceptor and ConnectInterceptor? (Or any other solution to override this behaviour on our side)

Impacted class: https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/RealCall.kt
```
@Throws(IOException::class)
  fun getResponseWithInterceptorChain(): Response {
    // Build a full stack of interceptors.
    val interceptors = mutableListOf<Interceptor>()
    interceptors += client.interceptors
    interceptors += RetryAndFollowUpInterceptor(client)
    interceptors += BridgeInterceptor(client.cookieJar)
    interceptors += CacheInterceptor(client.cache)
    interceptors += ConnectInterceptor
    if (!forWebSocket) {
      interceptors += client.networkInterceptors
    }
    interceptors += CallServerInterceptor(forWebSocket)

    val chain = RealInterceptorChain(interceptors, transmitter, null, 0, originalRequest, this,
        client.connectTimeoutMillis, client.readTimeoutMillis, client.writeTimeoutMillis)
```

**Comment by swankjesse** [comment]
Use a network interceptor to rewrite the returned Location header.

**Comment by yschimke** [comment]
@swankjesse do you know if Envoy works as a http proxy?

**Comment by swankjesse** [comment]
It works very well!

**Comment by yschimke** [comment]
> It works very well!

So can you just work on the world of your end targets, so no URLs pointing to the local proxy, and instead just configure a http proxy for any services within the domain of your service mesh?

**Comment by swankjesse** [comment]
Yep. We’re doing everything cleartext between application and envoy. We were using H2_PRIOR_KNOWLEDGE but we ran into the flow control window bugs that we fixed in 4.2.0.

**Comment by anycz** [comment]
Just tested it, works like a charm.
Thanks for the quick answer ;)
